runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -324,7 +324,12 @@ void J9::RecognizedCallTransformer::process_java_lang_StringUTF16_toBytes(TR::Tr
    newCallNode->setAndIncChild(3, TR::Node::iconst(0));
    newCallNode->setAndIncChild(4, lenNode);
 
-   treetop->insertAfter(TR::TreeTop::create(comp(), TR::Node::create(node, TR::treetop, 1, newCallNode)));
+   TR::Node *allocationFence = TR::Node::createAllocationFence(newByteArrayNode, newByteArrayNode);
+
+   TR::TreeTop* newTT = treetop->insertAfter(TR::TreeTop::create(comp(), TR::Node::create(node, TR::treetop, 1, newCallNode)));
+   // Insert the allocationFence after the arraycopy because the array can be allocated from the non-zeroed TLH
+   // and therefore we need to make sure no other thread sees stale memory from the array element section.
+   newTT->insertAfter(TR::TreeTop::create(comp(), allocationFence));
```

Comment:
Are allocation fence nodes generated in general after allocations on platforms such as X86, that do not have weak memory model ? I ask, because this insertion of an allocation fence seems to be unconditional in that sense, it will add it across platforms. This may not be incorrect, but I thought I'd check if it was unconventional (e.g. if no one code ever generated an allocation fence on non-weak ordered platforms) since that could lead to bugs in its own way (i.e. IL that is unexpectedly different).

Reply:
Yes, good catch.. I had looked at this but somehow missed the `cg()->getEnforceStoreOrder()` check in the genFence() routine used in ILGen. I added that check now.